src/client/mod.rs

```diff
@@ -483,37 +523,331 @@ impl IotHub for IotHubClient {
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx_twin_desired, mut rx_twin_desired) = mpsc::channel(100);
+    ///
+    ///     #[cfg(feature = "edge_client")]
+    ///     let mut client = IotHubClient::builder().observe_desired_properties(tx_twin_desired).build_edge_client().unwrap();
+    ///     #[cfg(feature = "device_client")]
+    ///     let mut client = IotHubClient::builder().observe_desired_properties(tx_twin_desired).build_device_client("my-connection-string").unwrap();
+    ///     #[cfg(feature = "module_client")]
+    ///     let mut client = IotHubClient::builder().observe_desired_properties(tx_twin_desired).build_module_client("my-connection-string").unwrap();
+    ///
+    ///     loop {
+    ///         select! (
+    ///             status = rx_twin_desired.recv() => {
+    ///                 // handle twin desired properties;
+    ///                 // ...
+    ///             },
+    ///         )
+    ///     }
+    /// }
+    /// ```
+    fn observe_desired_properties(
+        mut self: Box<Self>,
+        tx_twin_desired: mpsc::Sender<(TwinUpdateState, serde_json::Value)>,
+    ) -> Box<dyn IotHubBuilder> {
+        self.tx_twin_desired = Some(tx_twin_desired);
+        self
+    }
 
-        let mut client = Box::new(IotHubClient {
-            twin,
-            tx_connection_status,
-            tx_twin_desired,
-            tx_direct_method,
-            tx_incoming_message,
-            confirmation_set: JoinSet::new(),
+    /// Add direct method observer
+    /// ```no_run
+    /// use azure_iot_sdk::client::*;
+    /// use std::{thread, time};
+    /// use tokio::{select, sync::mpsc};
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx_direct_method, mut rx_direct_method) = mpsc::channel(100);
+    ///
+    ///     #[cfg(feature = "edge_client")]
+    ///     let mut client = IotHubClient::builder()
+    ///         .observe_direct_methods(tx_direct_method)
+    ///         .build_edge_client()
+    ///         .unwrap();
+    ///     #[cfg(feature = "device_client")]
+    ///     let mut client = IotHubClient::builder()
+    ///         .observe_direct_methods(tx_direct_method)
+    ///         .build_device_client("my-connection-string")
+    ///         .unwrap();
+    ///     #[cfg(feature = "module_client")]
+    ///     let mut client = IotHubClient::builder()
+    ///         .observe_direct_methods(tx_direct_method)
+    ///         .build_module_client("my-connection-string")
+    ///         .unwrap();
+    ///
+    ///     loop {
+    ///         select! (
+    ///             status = rx_direct_method.recv() => {
+    ///                 // handle direct method calls;
+    ///                 // ...
+    ///             },
+    ///         )
+    ///     }
+    /// }
+    /// ```
+    fn observe_direct_methods(
+        mut self: Box<Self>,
+        tx_direct_method: DirectMethodSender,
+    ) -> Box<dyn IotHubBuilder> {
```

Comment:
Why do we pass around the builder as a boxed type here? IMHO a mut& would be sufficient.

Reply:
since it is a trait type with unknown size

Reply:
How about using `Self` instead of the concrete type? Thus we erase the type of the trait in the return type. See: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1a88bd69a76be35533a302526aad33eb